sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -318,6 +318,11 @@ object Cast extends QueryErrorsBase {
     case _ if from == to => true
     case (NullType, _) => true
     case (_: NumericType, _: NumericType) => true
+    case (_: AtomicType, CharType(_) | VarcharType(_))
+      if !SQLConf.get.preserveCharVarcharTypeInfo => false
+    case (CharType(a), CharType(b)) => a <= b
+    case (CharType(a), VarcharType(b)) => a <= b
+    case (VarcharType(a), VarcharType(b)) => a <= b
```

Comment:
Is this change necessary in this PR? I would say opening a separate PR to fix casting is better than mixing stuff.